awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Graph.jsx

```diff
@@ -0,0 +1,756 @@
+
+    g.nodes().forEach(node => {
+      nodePositions[node] = g.node(node);
+    });
+
+    const linkRefs = svgGroup
+      .selectAll('.WorkflowChart-link')
+      .data(links, d => {
+        return `${d.source.id}-${d.target.id}`;
+      });
+
+    // Remove any stale links
+    linkRefs.exit().remove();
+
+    // Add any new links
+    const linkEnter = linkRefs
+      .enter()
+      .append('g')
+      .attr('class', 'WorkflowChart-link')
+      .attr('id', d => `link-${d.source.id}-${d.target.id}`)
+      .attr('stroke-width', '2px');
+
+    linkEnter
+      .append('polygon', 'g')
+      .attr('class', 'WorkflowChart-linkOverlay')
+      .attr('fill', '#E1E1E1')
+      .style('opacity', '0')
+      .attr('id', d => `link-${d.source.id}-${d.target.id}-overlay`)
+      .attr('points', d => getLinkOverlayPoints(d))
+      .on('mouseenter', d => {
+        setHelpText(d);
+      })
+      .on('mouseleave', () => {
+        setHelpText(null);
+      });
+
+    // Add entering links in the parent’s old position.
+    linkEnter
+      .insert('path', 'g')
+      .attr('class', 'WorkflowChart-linkPath')
+      .attr('d', d => lineData(d))
+      .attr('stroke', d => {
+        if (d.edgeType) {
+          if (d.edgeType === 'failure') {
+            return '#d9534f';
+          }
+          if (d.edgeType === 'success') {
+            return '#5cb85c';
+          }
+          if (d.edgeType === 'always') {
+            return '#337ab7';
+          }
+        }
+        return '#D7D7D7';
+      });
+
+    linkEnter
+      .append('polygon', 'g')
+      .style('opacity', '0')
+      .attr('points', d => getLinkOverlayPoints(d))
+      .on('mouseenter', d => {
+        setHelpText(d);
+      })
+      .on('mouseleave', () => {
+        setHelpText(null);
+      });
+
+    linkEnter
+      .on('mouseenter', d => {
+        d3.select(`#link-${d.source.id}-${d.target.id}`).raise();
+        d3.select(`#link-${d.source.id}-${d.target.id}-overlay`).style(
+          'opacity',
+          '1'
+        );
+        if (!readOnly) {
+          d3
+            .select(`#link-${d.source.id}-${d.target.id}`)
+            .append('foreignObject')
+            .attr('transform', () => {
+              const normalizedSourceY = normalizeY(
+                nodePositions,
+                nodePositions[d.source.id].y
+              );
+              const halfSourceHeight = nodePositions[d.source.id].height / 2;
+              const normalizedTargetY = normalizeY(
+                nodePositions,
+                nodePositions[d.target.id].y
+              );
+              const halfTargetHeight = nodePositions[d.target.id].height / 2;
+
+              let yPos =
+                (normalizedSourceY +
+                  halfSourceHeight +
+                  normalizedTargetY +
+                  halfTargetHeight) /
+                2;
+
+              if (d.source.id === 1) {
+                yPos += 4;
+              }
+
+              yPos -= 34;
```

Comment:
These numbers seem pretty significant. Perhaps they should be constants / variables?